packages/runtime/container-runtime/src/runningSummarizer.ts

```diff
@@ -309,38 +321,99 @@ export class RunningSummarizer implements IDisposable {
         })().finally(() => {
             summarizingLock.resolve();
             this.summarizingLock = undefined;
-            this.checkRerunHeuristics();
+            this.checkSummarizeAgain();
         }).catch((error) => {
             this.logger.sendErrorEvent({ eventName: "UnexpectedSummarizeError" }, error);
         });
     }
 
     public summarizeOnDemand(
         reason: string,
-        options: Omit<ISubmitSummaryOptions, "summaryLogger">,
+        { afterSequenceNumber, ...options }: IOnDemandSummarizeOptions,
     ): OnDemandSummarizeResult {
-        // Check for concurrent summary attempts. If one is found,
-        // return a promise that caller can await before trying again.
-        if (this.summarizingLock !== undefined) {
-            // The heuristics are blocking concurrent summarize attempts.
-            return { alreadyRunning: this.summarizingLock };
+        if (this.stopping || this.disposed) {
+            const failBuilder = new SummarizeResultBuilder();
+            failBuilder.fail("RunningSummarizer stopped or disposed", undefined);
+            return failBuilder.build();
         }
-        if (this.generator.isSummarizing()) {
-            // Another summary is currently being generated.
-            return { alreadyRunning: this.generator.waitSummarizing() };
+        const onDemandReason = `onDemand;${reason}` as const;
+        if (afterSequenceNumber === undefined) {
+            // Check for concurrent summary attempts. If one is found,
+            // return a promise that caller can await before trying again.
+            if (this.summarizingLock !== undefined) {
+                // The heuristics are blocking concurrent summarize attempts.
+                return { alreadyRunning: this.summarizingLock };
+            }
+            if (this.generator.isSummarizing()) {
+                // Another summary is currently being generated.
+                return { alreadyRunning: this.generator.waitSummarizing() };
+            }
+            const result = this.generator.summarize(onDemandReason, options);
+            result.receivedSummaryAckOrNack.finally(() => this.checkSummarizeAgain());
+            return result;
+        }
+        // If afterSequenceNumber is provided, the on-demand summary is "enqueued" to run
+        // once an eligible op comes in with sequenceNumber >= afterSequenceNumber.
+        if (this.enqueuedSummary !== undefined) {
+            // Note that it could be already generating a summary at this point in time.
```

Comment:
For simplicity, you could do the `summarizingLock` and `isSummarizing` check irrespective of whether `afterSequenceNumber` is defined or not. So, if summarize is in progress, the caller should try again later.

Reply:
Another option is to always queue the summary irrespective of whether `afterSequenceNumber` is defined or not. That summary will run once the current one completes.
I think we should keep both these flows consistent.

Reply:
For the first comment, I can't do the check against the locks for both cases, because when afterSequenceNumber is provided, it may be for a much later sequence number.

Reply:
For the second comment, I was thinking both methods of on-demand summarization could be useful. Maybe I can change the signatures to better reflect the intent?

When you would want enqueue: need to ensure a summary occurs by a specific sequence number. Not worried about excess summary attempts.

When you would want existing behavior: running your own heuristics, and want to make sure a summary attempts within reasonable time. If one is already going, maybe that's sufficient and you would want to recompute your heuristics once the alreadyRunning promise resolves.

Reply:
After discussion, I split into 2 functions:
1. summarizeOnDemand - current behavior; if a summary is already in progress, it will just return the alreadyRunning promise, otherwise it will summarize immediately.
2. enqueueSummarize - always will try to summarize at next opportunity, but if afterSequenceNumber is provided, then it will wait until after that sequence number elapses.